hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/HazelcastSqlCaseOperator.java

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.operators;
+
+import com.hazelcast.sql.impl.calcite.validate.HazelcastSqlValidator;
+import com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlCase;
+import org.apache.calcite.sql.fun.SqlCaseOperator;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.InferTypes;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Custom Hazelcast {@link SqlCaseOperator} to override the default return type
+ * inference strategy for CASE.
+ *
+ * @see HazelcastSqlCase
+ */
+public final class HazelcastSqlCaseOperator extends SqlOperator {
+
+    public HazelcastSqlCaseOperator() {
+        super(SqlCaseOperator.INSTANCE.getName(), SqlKind.CASE, MDX_PRECEDENCE, true, null, InferTypes.RETURN_TYPE, null);
+    }
+
+    @Override
+    public void validateCall(SqlCall call, SqlValidator validator, SqlValidatorScope scope, SqlValidatorScope operandScope) {
+        SqlCaseOperator.INSTANCE.validateCall(call, validator, scope, operandScope);
+    }
+
+    @Override
+    public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+        // SqlCaseOperator is doing the same
+        return validateOperands(validator, scope, call);
+    }
+
+    @Override
+    public boolean checkOperandTypes(SqlCallBinding binding, boolean throwOnFailure) {
+        return SqlCaseOperator.INSTANCE.checkOperandTypes(binding, throwOnFailure);
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding binding) {
+        if (binding instanceof SqlCallBinding) {
+            SqlCallBinding sqlBinding = (SqlCallBinding) binding;
+            SqlCase call = (SqlCase) sqlBinding.getCall();
+            HazelcastSqlValidator validator = (HazelcastSqlValidator) sqlBinding.getValidator();
+
+            validator.getTypeCoercion().caseWhenCoercion(sqlBinding);
+            RelDataType caseType = validator.getKnownNodeType(call);
+            if (caseType == null) {
+                throw sqlBinding.newValidationError(RESOURCE.dynamicParamIllegal());
```

Comment:
Calcite's error message is too vague here. What exactly is wrong with the dynamic parameter in this branch?

Reply:
Resolving. There is a ticket to clarify parameter behavior.